---
title: "Command-line use of bioinformatics tools"
author: "Jacques van Helden"
date: 'Created: 8/10/2018; last update: `r Sys.Date()`'
output:
  html_document:
    self_contained: false
    fig_caption: yes
    highlight: zenburn
    theme: cerulean
    toc: yes
    toc_depth: 3
    toc_float: yes
  md_document:
    variant: markdown_github
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 2
  word_document:
    toc: yes
    toc_depth: 2
#bibliography: ./geneset_functional_analysis.bib
---


```{r setup, include=FALSE}
library(knitr, warn.conflicts = FALSE)
knitr::opts_chunk$set(
  echo = TRUE, 
  results = TRUE, 
  message = FALSE,
  fig.align = "center")

dir.main <- "~/IFB/NNCR/using_IFB_NNCR/"
dir.tp <- file.path(dir.main, "practicals", "")
setwd(dir.tp)

```

****************************************************************
# Prerequisite

This practical assumes that you are connected to a cluster of the IFB NNCR. This enables you to benefit from a software environment with pre-installed software tools for the different fields of bioinformatics.

To open a connection to the IFB NNCR, follow the totirla  [Tutorial](../NNCR-cluster_connection/NNCR-cluster_connection.html) | 


# Resources

| Resource | URL and description |
|-----------------------|--------------------------------------------|
| Ensembl Bacteria | https://bacteria.ensembl.org/index.html <br>Database and browser for bacterial and archaeal genomes maintained at the European Bioinformatics Institute (EBI) |
| GTF format | <https://genome.ucsc.edu/FAQ/FAQformat.html#format4> <br>Specification of the Gene Transfer Format (GTF) on the UCSC genome browser Web site | 

# Tutorial

## Browsing genomes with EnsemblGenomes

As reference genome, we will use the strain *K12 MG1655* of  *Escherichia coli*. 

1. In a web browser, open a connexion to Ensembl Bacteria (<https://bacteria.ensembl.org>).

2. In the  **Search for a genome** box, type *MG1655*. The interface will display a list of the genomes matching this string. Select the strain with taxonomic ID (TAXID) *511145*.

```{r out.width="60%", fig.cap="**EnsemblGenomes Bacteria organism choice**. ", echo=FALSE}
include_graphics("images/ensembl_bacteria_organism_selection.png")
```
3. After having selected your bacteria of interest, type *aspartokinase* in the search box and start the search. 

```{r out.width="60%", fig.cap="**EnsemblGenomes gene search**. ", echo=FALSE}
include_graphics("images/ensembl_bacteria_search_aspartokinase.png")
```


4. Read the annotation of the three first matches. Note the names of the genes coding for these three enzymes. 

```{r out.width="60%", fig.cap="**EnsemblGenomes search result for *aspartokinase***. ", echo=FALSE}
include_graphics("images/Ecoli_aspartokinase_search_result.png")
```

## Downloading genomic features and protein sequences

We will use blast on the command line to search for homologs of *E.coli*  aspartokinases in other bacterial genomes.

Each student will chose a different bacteria to perform the search. 

## Downloading protein sequences and gene annotations from Ensembl genome

1. In a separate window, open a connection to [EnsemblGenomes Bacteria](https://bacteria.ensembl.org/) (<https://bacteria.ensembl.org/>).

2. Click on the **Downloads* link (you can also access it directly here: <https://bacteria.ensembl.org/info/website/ftp/>). 

3. In the **filter** box, type MG1655. This select the subset of genomes whose name contains this string. 

```{r out.width="60%", fig.cap="**EnsemblGenomes download filter**. ", echo=FALSE}
include_graphics("images/ensemblgenomes_bacteria_ftp_mg1655.png")
```

4. Click on the **FASTA** link in the column **Protein sequences**/ 


5. and select **Copy link address**. 

```{r out.width="200", fig.cap="**EnsemblGenomes** Link to download all protein sequences for a selected organism. ", echo=FALSE}
include_graphics("images/protein_fasta_link.png")
```

5. Come back to the Unix terminal of the IFB NNCR.

6. Create a directory for the current practical. 

```{bash eval=FALSE}
mkdir -p ~/bacteria_blast
cd ~/bacteria_blast
```


7. type `wget -rNL` and paste the content of the clipboard. This should give the following command (note: at the tiem you run this tutorial, the release may be different, this one is from October 2018). 

```{bash eval=FALSE}
wget --no-clobber ftp://ftp.ensemblgenomes.org/pub/release-41/bacteria//fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/pep/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa.gz

## Check that the file is there
ls -l

## Uncompress the file
gunzip Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa.gz

# Measure the size of the uncompressed file
du -k Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa

# Check the 20 fist lines of the file
head -n 20 Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa

# Check the 20 last lines of the file
tail -n 20 Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa
```


## Downloading genome annotations

After having downloaded the protein sequences, we will download gene annotations. In EnsemblGenomes, annotations are available in different formats, including GTF, a text file with columns separated by tabulations. 

1. We In the browser, come back to the EnsemblGenomes Bacteria Download site, select MG1655 and follow the same instructions as above to download and uncompress the GTF file. 

2. Uncompress the gzi archive. 


## Exercise: genome annotations

As an exercise, you will now elaborate  Unix commands to explore the content of E.coli genome annotations. 

For this, we recommmend to use and combine the following commands. In order to fine-tune the parameters, read their manual on the Unix terminal by typing `man [command_name]` (for example `man wc`).

| Command | Usage |
|---------|---------------------------------------------------------|
| wc | count lines, words and characters of a file |
| cut | cut specific columns of a text file |
| grep | select rows of a file that match a given query (string, regular expression) |
| awk | a powerful tool to work on text files, combining row filters, column selections, and many other features. |
| sort | sort the rows of a text file based on user-specified columns. The sorting can be done either alphabetically or numerically, in ascending or descending order |
| uniq | discard redundant lines from a text file. The option `-c` counts the number of occurrences of each line in a redundant file |

You can learn more Unix commands by typing "Linux cheat sheet" in any Web search engine. 


We ask you to write all the answers in a text file, with comments, in order to enable other people to reproduce all your commands. 

### Questions

1. How many rows contains the GTF file for E.coli K12 MG1655 ?
2. How many annotation rows are there in the GTF file (for this, you need to discard the comment rows starting with a `#` character) ?

3. How many of these rows correspond to genes ? 

4. Count the number of instnces of each feature type in the GTF file. 

5. Select all the rows of the GTF file corresponding to a gene, and store them in a separate file named `genes_Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf`. 

```{bash eval=FALSE, echo=FALSE}
## Solution
wget --no-clobber ftp://ftp.ensemblgenomes.org/pub/release-41/bacteria//gtf/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf.gz

gunzip -f Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf.gz

## How many rows contains the GTF file for E.coli K12 MG1655 ?
wc -l Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf

## How many of these rows correspond to genes ? 
## How many annotation rows contains the GTF file 
grep -v '^#' Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf | wc -l

## How many of these rows correspond to genes ? 
grep -v '^#' Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf \
  | cut -f 3 | grep '^gene$' | wc -l

## Count the number of instnces of each feature type in the GTF file. 
grep -v '^#' Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf \
  | cut -f 3 \
  | sort \
  | uniq -c \
  | sort -n -r -k 1

## Select all the rows of the GTF file corresponding to a gene, and store them in a separate file named
grep -v '^#' Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf \
  | awk -F '\t' '$3 == "gene"' \
  > genes_Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf

## Check the number of rows selected by awk
wc -l genes_Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf



```


# Running blast for all proteins of a genome


## Activating the conda environment for blast on the NNCR cluster

In addition to facilitating the installation of software tools on different operating system, conda also permits to create modules that regroup a set of tools usually used together (for example NGS analysis, protein structure, proteomics, ...). 

The IFB NNCR cluster is already equipped with a series of conda environments with the classical bioinformatics tools. 

Before being able to use a tool, you must activate the corresponding environment. 

```{bash eval=FALSE}
## List the conda environments available on the system
$CONDA_HOME/bin/conda env list

## FInd the blast environment
$CONDA_HOME/bin/conda env list | grep -i blast

## Activate the blast environment. 
source $CONDA_HOME/bin/activate blast-2.7.1

## Check the location of the blastp command
which blastp

## Deactivate the current environment
source $CONDA_HOME/bin/deactivate

## Re-activate the blast environment
source $CONDA_HOME/bin/activate blast-2.7.1
```


## Formatting the blast database for your reference genome

```{bash eval=FALSE}
## Get the usage of makeblastdb (formal specification of the command syntax)
makeblastdb -h

## Get the help for makeblastdb (explanation of the options)
makeblastdb -help

## Buil blast database with all the protein sequences of E.coli
makeblastdb -in Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa  -dbtype prot
```

Get the help for *blastp*. 

```{bash eval=FALSE}
## Command  syntax
blastp -h

## Help
blastp -help

```

## E. coli versus E. coli

**Exercise: ** build a command that compares the sequences of all the E.coli proteins against each other, and store the results in a tab-delimited file with a single line per hit (we do not need the alignments).   


****************************************************************
# References







