<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Jacques van Helden" />


<title>Using BLAST on the command line</title>

<script src="blast_proteome_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="blast_proteome_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="blast_proteome_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="blast_proteome_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="blast_proteome_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="blast_proteome_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="blast_proteome_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="blast_proteome_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="blast_proteome_files/navigation-1.1/tabsets.js"></script>
<script src="blast_proteome_files/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #303030; color: #cccccc; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; }
td.sourceCode { padding-left: 5px; }
pre, code { color: #cccccc; background-color: #303030; }
code > span.kw { color: #f0dfaf; } /* Keyword */
code > span.dt { color: #dfdfbf; } /* DataType */
code > span.dv { color: #dcdccc; } /* DecVal */
code > span.bn { color: #dca3a3; } /* BaseN */
code > span.fl { color: #c0bed1; } /* Float */
code > span.ch { color: #dca3a3; } /* Char */
code > span.st { color: #cc9393; } /* String */
code > span.co { color: #7f9f7f; } /* Comment */
code > span.ot { color: #efef8f; } /* Other */
code > span.al { color: #ffcfaf; } /* Alert */
code > span.fu { color: #efef8f; } /* Function */
code > span.er { color: #c3bf9f; } /* Error */
code > span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */
code > span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code > span.sc { color: #dca3a3; } /* SpecialChar */
code > span.vs { color: #cc9393; } /* VerbatimString */
code > span.ss { color: #cc9393; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { } /* Variable */
code > span.cf { color: #f0dfaf; } /* ControlFlow */
code > span.op { color: #f0efd0; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code > span.at { } /* Attribute */
code > span.do { color: #7f9f7f; } /* Documentation */
code > span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code > span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code > span.in { color: #7f9f7f; font-weight: bold; } /* Information */
div.sourceCode {
  overflow-x: visible;
}
</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>



<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Using BLAST on the command line</h1>
<h4 class="author"><em>Jacques van Helden</em></h4>
<h4 class="date"><em>Created: 8/10/2018; last update: 2018-10-11</em></h4>

</div>


<hr />
<div id="prerequisite" class="section level1">
<h1>Prerequisite</h1>
<p>This practical assumes that you are connected to a cluster of the IFB NNCR. This enables you to benefit from a software environment with pre-installed software tools for the different fields of bioinformatics.</p>
<p>To open a connection to the IFB NNCR, follow the totirla <a href="../NNCR-cluster_connection/NNCR-cluster_connection.html">Tutorial</a> |</p>
</div>
<div id="resources" class="section level1">
<h1>Resources</h1>
<table style="width:96%;">
<colgroup>
<col width="33%" />
<col width="62%" />
</colgroup>
<thead>
<tr class="header">
<th>Resource</th>
<th>URL and description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Ensembl Bacteria</td>
<td><a href="https://bacteria.ensembl.org/index.html" class="uri">https://bacteria.ensembl.org/index.html</a> <br>Database and browser for bacterial and archaeal genomes maintained at the European Bioinformatics Institute (EBI)</td>
</tr>
<tr class="even">
<td>GTF format</td>
<td><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format4" class="uri">https://genome.ucsc.edu/FAQ/FAQformat.html#format4</a> <br>Specification of the Gene Transfer Format (GTF) on the UCSC genome browser Web site</td>
</tr>
<tr class="odd">
<td>Ensembl Bacteria Metadata</td>
<td><a href="ftp://ftp.ensemblgenomes.org/pub/release-41/bacteria/species_EnsemblBacteria.txt" class="uri">ftp://ftp.ensemblgenomes.org/pub/release-41/bacteria/species_EnsemblBacteria.txt</a><br>table with information about each sequenced genome</td>
</tr>
<tr class="even">
<td>Selected genomes</td>
<td><a href="data/microme_selected_organisms_cross_references.tsv">MICROME genome table</a><br>A table containing 80 genomes of interest for the MICROME European network.</td>
</tr>
</tbody>
</table>
</div>
<div id="tutorial" class="section level1">
<h1>Tutorial</h1>
<div id="browsing-genomes-with-ensemblgenomes" class="section level2">
<h2>Browsing genomes with EnsemblGenomes</h2>
<p>As reference genome, we will use the strain <em>K12 MG1655</em> of <em>Escherichia coli</em>.</p>
<ol style="list-style-type: decimal">
<li><p>In a web browser, open a connexion to Ensembl Bacteria (<a href="https://bacteria.ensembl.org" class="uri">https://bacteria.ensembl.org</a>).</p></li>
<li><p>In the <strong>Search for a genome</strong> box, type <em>MG1655</em>. The interface will display a list of the genomes matching this string. Select the strain with taxonomic ID (TAXID) <em>511145</em>.</p></li>
</ol>
<div class="figure" style="text-align: center">
<img src="images/ensembl_bacteria_organism_selection.png" alt="**EnsemblGenomes Bacteria organism choice**. " width="60%" />
<p class="caption">
<strong>EnsemblGenomes Bacteria organism choice</strong>.
</p>
</div>
<ol start="3" style="list-style-type: decimal">
<li>After having selected your bacteria of interest, type <em>aspartokinase</em> in the search box and start the search.</li>
</ol>
<div class="figure" style="text-align: center">
<img src="images/ensembl_bacteria_search_aspartokinase.png" alt="**EnsemblGenomes gene search**. " width="60%" />
<p class="caption">
<strong>EnsemblGenomes gene search</strong>.
</p>
</div>
<ol start="4" style="list-style-type: decimal">
<li>Read the annotation of the three first matches. Note the names of the genes coding for these three enzymes.</li>
</ol>
<div class="figure" style="text-align: center">
<img src="images/Ecoli_aspartokinase_search_result.png" alt="**EnsemblGenomes search result for *aspartokinase***. " width="60%" />
<p class="caption">
<strong>EnsemblGenomes search result for <em>aspartokinase</em></strong>.
</p>
</div>
</div>
<div id="downloading-genomic-features-and-protein-sequences" class="section level2">
<h2>Downloading genomic features and protein sequences</h2>
<p>We will use blast on the command line to search for homologs of <em>E.coli</em> aspartokinases in other bacterial genomes.</p>
<p>Each student will chose a different bacteria to perform the search.</p>
</div>
<div id="downloading-protein-sequences-and-gene-annotations-from-ensembl-genome" class="section level2">
<h2>Downloading protein sequences and gene annotations from Ensembl genome</h2>
<ol style="list-style-type: decimal">
<li><p>In a separate window, open a connection to <a href="https://bacteria.ensembl.org/">EnsemblGenomes Bacteria</a> (<a href="https://bacteria.ensembl.org/" class="uri">https://bacteria.ensembl.org/</a>).</p></li>
<li><p>Click on the **Downloads* link (you can also access it directly here: <a href="https://bacteria.ensembl.org/info/website/ftp/" class="uri">https://bacteria.ensembl.org/info/website/ftp/</a>).</p></li>
<li><p>In the <strong>filter</strong> box, type MG1655. This select the subset of genomes whose name contains this string.</p></li>
</ol>
<div class="figure" style="text-align: center">
<img src="images/ensemblgenomes_bacteria_ftp_mg1655.png" alt="**EnsemblGenomes download filter**. " width="60%" />
<p class="caption">
<strong>EnsemblGenomes download filter</strong>.
</p>
</div>
<ol start="4" style="list-style-type: decimal">
<li><p>Click on the <strong>FASTA</strong> link in the column <strong>Protein sequences</strong>/</p></li>
<li><p>and select <strong>Copy link address</strong>.</p></li>
</ol>
<div class="figure" style="text-align: center">
<img src="images/protein_fasta_link.png" alt="**EnsemblGenomes** Link to download all protein sequences for a selected organism. " width="200" />
<p class="caption">
<strong>EnsemblGenomes</strong> Link to download all protein sequences for a selected organism.
</p>
</div>
<ol start="5" style="list-style-type: decimal">
<li><p>Come back to the Unix terminal of the IFB NNCR.</p></li>
<li><p>Create a directory for the current practical.</p></li>
</ol>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> -p ~/bacteria_blast
<span class="bu">cd</span> ~/bacteria_blast</code></pre></div>
<ol start="7" style="list-style-type: decimal">
<li>type <code>wget --no-clubber</code> and paste the content of the clipboard. This should give the following command (note: at the tiem you run this tutorial, the release may be different, this one is from October 2018).</li>
</ol>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">wget</span> --no-clobber ftp://ftp.ensemblgenomes.org/pub/release-41/bacteria//fasta/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/pep/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa.gz</code></pre></div>
<p>If you want to know the meaning of the <code>--no-clobber</code> option you can checek <code>wget</code> manual with the command.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">man</span> wget</code></pre></div>
<p>We will now check if the file has been correctly downloaded, and uncompress it.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">
<span class="co">## Check that the file is there</span>
<span class="fu">ls</span> -l

<span class="co">## Uncompress the file</span>
<span class="fu">gunzip</span> Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa.gz

<span class="co"># Measure the size of the uncompressed file</span>
<span class="fu">du</span> -k Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa

<span class="co"># Check the 20 fist lines of the file</span>
<span class="fu">head</span> -n 20 Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa

<span class="co"># Check the 20 last lines of the file</span>
<span class="fu">tail</span> -n 20 Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa</code></pre></div>
</div>
<div id="downloading-genome-annotations" class="section level2">
<h2>Downloading genome annotations</h2>
<p>After having downloaded the protein sequences, we will download gene annotations. In EnsemblGenomes, annotations are available in different formats, including GTF, a text file with columns separated by tabulations.</p>
<ol style="list-style-type: decimal">
<li><p>We In the browser, come back to the EnsemblGenomes Bacteria Download site, select MG1655 and follow the same instructions as above to download and uncompress the GTF file.</p></li>
<li><p>Uncompress the gzip archive.</p></li>
</ol>
<div id="solution" class="section level3">
<h3>Solution</h3>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## Download genome annotation file (GTF format)</span>
<span class="fu">wget</span> --no-clobber ftp://ftp.ensemblgenomes.org/pub/release-41/bacteria//gtf/bacteria_0_collection/escherichia_coli_str_k_12_substr_mg1655/Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf.gz

<span class="co">## Uncompress genome annotation file</span>
<span class="fu">gunzip</span> Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf.gz

<span class="co">## Check that the uncompressed file is there</span>
<span class="fu">ls</span> Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf

<span class="co">## Check the head of the annotaiton file</span>
<span class="fu">head</span> -n 20 Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf

<span class="co">## Check the tail</span>
<span class="fu">tail</span> -n 10 Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf</code></pre></div>
</div>
</div>
<div id="exercise-genome-annotations" class="section level2">
<h2>Exercise: genome annotations</h2>
<p>As an exercise, you will now elaborate Unix commands to explore the content of E.coli genome annotations.</p>
<p>For this, we recommmend to use and combine the following commands. In order to fine-tune the parameters, read their manual on the Unix terminal by typing <code>man [command_name]</code> (for example <code>man wc</code>).</p>
<table style="width:94%;">
<colgroup>
<col width="13%" />
<col width="80%" />
</colgroup>
<thead>
<tr class="header">
<th>Command</th>
<th>Usage</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>wc</td>
<td>count lines, words and characters of a file</td>
</tr>
<tr class="even">
<td>cut</td>
<td>cut specific columns of a text file</td>
</tr>
<tr class="odd">
<td>grep</td>
<td>select rows of a file that match a given query (string, regular expression)</td>
</tr>
<tr class="even">
<td>awk</td>
<td>a powerful tool to work on text files, combining row filters, column selections, and many other features.</td>
</tr>
<tr class="odd">
<td>sort</td>
<td>sort the rows of a text file based on user-specified columns. The sorting can be done either alphabetically or numerically, in ascending or descending order</td>
</tr>
<tr class="even">
<td>uniq</td>
<td>discard redundant lines from a text file. The option <code>-c</code> counts the number of occurrences of each line in a redundant file</td>
</tr>
</tbody>
</table>
<p>You can learn more Unix commands by typing “Linux cheat sheet” in any Web search engine.</p>
<p>We ask you to write all the answers in a text file, with comments, in order to enable other people to reproduce all your commands.</p>
<div id="questions" class="section level3">
<h3>Questions</h3>
<ol style="list-style-type: decimal">
<li>How many rows contains the GTF file for E.coli K12 MG1655 ?</li>
<li><p>How many annotation rows are there in the GTF file (for this, you need to discard the comment rows starting with a <code>#</code> character) ?</p></li>
<li><p>How many of these rows correspond to genes ?</p></li>
<li><p>Count the number of instnces of each feature type in the GTF file.</p></li>
<li><p>Select all the rows of the GTF file corresponding to a gene, and store them in a separate file named <code>genes_Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.41.gtf</code>.</p></li>
</ol>
</div>
<div id="solution-1" class="section level3">
<h3>Solution</h3>
<hr />
</div>
</div>
</div>
<div id="searching-sequence-similarities-with-blast" class="section level1">
<h1>Searching sequence similarities with BLAST</h1>
<div id="activating-the-conda-environment-for-blast-on-the-nncr-cluster" class="section level2">
<h2>Activating the conda environment for blast on the NNCR cluster</h2>
<p>In addition to facilitating the installation of software tools on different operating system, conda also permits to create modules that regroup a set of tools usually used together (for example NGS analysis, protein structure, proteomics, …).</p>
<p>The IFB NNCR cluster is already equipped with a series of conda environments with the classical bioinformatics tools. Before being able to use a tool, you must activate the corresponding environment.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## List the conda environments available on the system</span>
<span class="va">$CONDA_HOME</span><span class="ex">/bin/conda</span> env list

<span class="co">## FInd the blast environment</span>
<span class="va">$CONDA_HOME</span><span class="ex">/bin/conda</span> env list <span class="kw">|</span> <span class="fu">grep</span> -i blast

<span class="co">## Activate the blast environment. </span>
<span class="bu">source</span> <span class="va">$CONDA_HOME</span>/bin/activate blast-2.7.1

<span class="co">## Check the location of the blastp command</span>
<span class="fu">which</span> blastp

<span class="co">## Deactivate the current environment. </span>
<span class="bu">source</span> <span class="va">$CONDA_HOME</span>/bin/deactivate
<span class="co">## Note the change in the shell prompt</span>

<span class="co">## Re-activate the blast environment</span>
<span class="bu">source</span> <span class="va">$CONDA_HOME</span>/bin/activate blast-2.7.1
<span class="co">## Note the change in the shell prompt</span></code></pre></div>
<p>Note: if you are working on a machine that does not belong to the federation of NNCR clusters, you can create the blast environment in your own conda folder with the following command (assuming that the conda package has been installed and that its executable files are in your path).</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## </span><span class="al">ATTENTION</span><span class="co">: DO NOT RUN THIS COMMAND IF YOU ARE ON THE NNCR CLUSTER,</span>
<span class="co">## because this would occupy a big space to reinstall BLAST in your </span>
<span class="co">## own account, although it is already available for all users. </span>

<span class="co">## Install blast</span>
<span class="ex">conda</span> create -n blast-2.7.1 blast==2.7.1

<span class="co">## Install emboss</span>
<span class="ex">conda</span> create -n emboss-6.6.0 emboss==6.6.0</code></pre></div>
</div>
<div id="formatting-the-blast-database-for-the-reference-genome" class="section level2">
<h2>Formatting the blast database for the reference genome</h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## Check the path of the program makeblastdb (it should be in your conda folder)</span>
<span class="fu">which</span> makeblastdb

<span class="co">## Get the usage of makeblastdb (formal specification of the command syntax)</span>
<span class="ex">makeblastdb</span> -h

<span class="co">## Get the help for makeblastdb (explanation of the options)</span>
<span class="ex">makeblastdb</span> -help

<span class="co">## Build a BLAST database with all the protein sequences of E.coli</span>
<span class="ex">makeblastdb</span> -in Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa  -dbtype prot

<span class="co">## Check the new files that were created with this commands. </span>
<span class="co">## For this we list the files in reverse (-r) temporal (-t) order. </span>
<span class="fu">ls</span> -1tr</code></pre></div>
<p>The program <code>makeblastdb</code> created three files besides the input fasta file.</p>
<pre><code>Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa.pin
Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa.phr
Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa.psq</code></pre>
<p>These files contain an index of all the oligopeptites found in all the sequences of the fasta file. This “k-mer” index will be used by BLAST to rapidly find all the sequences in the database that match a query sequence, and use these short correspondences to start an alignment.</p>
</div>
<div id="read-blastp-manual" class="section level2">
<h2>Read <em>blastp</em> manual</h2>
<p>Since we want to search a protein database with protein query sequences, we will use the <code>blastp</code> command.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## Command  syntax</span>
<span class="ex">blastp</span> -h

<span class="co">## Help</span>
<span class="ex">blastp</span> -help</code></pre></div>
<p>OK, the first contact is a bit frightening, because blastp has many options. This is because this program allows you to run queries in different modes, with different parameters, and to export the results in different formats.</p>
<p>In this tutorial we will show you the most usual ways to use the tool, and you will then be able to refine your search by combing back to the manual and understanding the use of additional options.</p>
</div>
<div id="downloading-a-query-protein" class="section level2">
<h2>Downloading a query protein</h2>
<p>We will first search similarities from a given protein of interest. We will start from the “askartokinase 3” protein from <em>Escherichia coli</em>, which is the product of the gene lysC.</p>
<p>Before starting the analysis, read the Uniprot record (<a href="https://www.uniprot.org/uniprot/P08660" class="uri">https://www.uniprot.org/uniprot/P08660</a>) in order to know the function and regulation of this protein.</p>
<p>We will use BLAST to identify all the proteins of <em>Escherichia coli</em> that have significant similarities with the aspartokinase 3.</p>
<p>For this we first need to download the sequence of this protein on our computer. The sequence can be downloaded directly from Uniprot.</p>
<pre class="blast"><code>wget --no-clobber https://www.uniprot.org/uniprot/P08660.fasta</code></pre>
</div>
<div id="similiraty-search-for-one-sequence-against-a-database" class="section level2">
<h2>Similiraty search for one sequence against a database</h2>
<p>The following command search similarities for one or several query sequences provided in a fasta file (<code>P08660.fasta</code>) in a protein sequence database (<code>Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa</code>). By default, blastp prints the result on the screen, but we prefer to redirect it to a file in order to keep a trace of the result and inspect it later.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## Search similarities for aspartokinase 3 in E.coli proteome</span>
<span class="ex">blastp</span> \
  -db Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa  \
  -query P08660.fasta \
  <span class="op">&gt;</span> lysC_vs_Ecoli.txt
  
<span class="co">## Check the result</span>
<span class="fu">less</span> lysC_vs_Ecoli.txt</code></pre></div>
<div id="question-set-1" class="section level3">
<h3>Question set 1</h3>
<p><font color="#004488"> 1. How many hits were found in total? 2. How many of these have an E-value higher than 1? 3. What was the threshold on the E-value for this blastp search? 4. With this threshold, how many hits would we expeect if we use a random sequence as query? 5. How many alignments would you consider significant in this result? 6. How many of these would you qualify of homologs? 7. For each of these putative homologs, is it a paralog, ortholog or “other log”? 8. Do you see a relationship between the E-value and some properties of the alignments? </font></p>
</div>
</div>
<div id="getting-a-tabular-synthesis-of-blast-results" class="section level2">
<h2>Getting a tabular synthesis of blast results</h2>
<p>By default, BLAST displays the detailed results of a similaty search, starting with a summary of the matches, followed by all the alignemnts. It can be convenient to generate a synthetic result indicating only the relevant statistics for each alignment. This can be done by tuning BLAST options.</p>
<p>Read blastp manual and find a way to tune the formatting options in order to get a tabular output with comments.</p>
<div id="solution-2" class="section level3">
<h3>Solution</h3>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">blastp</span> -outfmt 7 \
  -db Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa  \
  -query P08660.fasta \
  <span class="op">&gt;</span> lysC_vs_Ecoli_synthesis.tsv</code></pre></div>
</div>
</div>
</div>
<div id="negative-control-blast-against-a-shuffled-sequence" class="section level1">
<h1>Negative control: blast against a shuffled sequence</h1>
<p>In this section, we will perform a quick empirical test, by submitting a random sequence to blastp. For this, we will shuffle the aminoacids of the original query sequence.</p>
<div id="shuffling-sequences-with-a-web-tool" class="section level2">
<h2>Shuffling sequences with a Web tool</h2>
<p>One way to shuffle sequences is to use a convenient Web tool at <a href="http://www.bioinformatics.org/sms2/shuffle_protein.html" class="uri">http://www.bioinformatics.org/sms2/shuffle_protein.html</a></p>
<p>Here is an example of result (I just simplified the header of the shuffle output fasta sequence for the sake of readability).</p>
<p>Of course, your result should differ since each shuffling produces a randomized sequence.</p>
<pre><code>&gt;shuffled_lysC
VEAINNAVKRELSALGDVADLFVLEFDNDRLGWADAIIRVLLNTEMNGELKPAEVASAIV
GLIDEPELDTLVYSPLAGIFFLMCLIDFRGLGRREPTVPHSIMAGSGNALCRYFVTGLLL
LDTNAFAIRETKNLEAFVAEFQAADNFAQARNPANLPKLILCLSVNGVLQASPFFNLNEG
ADVVSESSAASLKADEDQSVITLEEDIVVTEGASSFDLALGLLMHLNVDRKVGAEWLEDE
VSPVMVEVLTLVLTQEALTVTSIALLRGDTRKRVGLVLNSFTSHDLIAPSLILIGTSLVT
RHFLVGAKVRIQERKASTSFRRLLAGRTETYRGLEIYAPCISMTTAVRRMQAGGIAHREI
QETIACRKSTELLSTIASPHAEVGFEAANGTLLGVLTADHVSMALQAVNGGESLPDEHRR
TVGDILTERVARLLLPFKEGPARTSSTLA</code></pre>
</div>
<div id="shuffling-sequences-with-emboss-shuffleseq" class="section level2">
<h2>Shuffling sequences with EMBOSS shuffleseq</h2>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## Activate the conda environment with the emboss software suite</span>
<span class="bu">source</span>  <span class="va">$CONDA_HOME</span>/bin/activate emboss-6.6.0

<span class="co">## Check that shuffleseq is available</span>
<span class="fu">which</span> shuffleseq

<span class="co">## Get the suffle manual </span>
<span class="ex">shuffleseq</span> -h

<span class="co">## Shuffle the apartokinase 3 sequence</span>
<span class="ex">shuffleseq</span> -sequence P08660.fasta -outseq shuffled_lysC.fa

<span class="co">## Check the result</span>
<span class="fu">cat</span> shuffled_lysC.fa

<span class="co">## Compare the shuffled seq with the original one</span>
<span class="fu">cat</span> P08660.fasta</code></pre></div>
</div>
<div id="exercise" class="section level2">
<h2>Exercise</h2>
<ol style="list-style-type: decimal">
<li><p>Shuffle the aspartokinase 3 sequence with the tool <code>shuffleseq</code> (found in the conda environment <code>emboss</code>).</p></li>
<li><p>Run a <code>blastp</code> search with this shuffled sequence against all proteins of <em>Escherichia coli</em>.</p></li>
</ol>
<div id="solution-3" class="section level3">
<h3>Solution</h3>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## Run blast with the shuffled lysC sequence against all E.coli proteins</span>
<span class="ex">blastp</span> \
  -db Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa  \
  -query shuffled_lysC.fa \
  <span class="op">&gt;</span> shuffled_lysC_vs_Ecoli.txt
<span class="co">## Check the result</span>
<span class="fu">less</span> shuffled_lysC_vs_Ecoli.txt

<span class="co">## Run the same query with a tabular output</span>
<span class="ex">blastp</span> -outfmt 7 \
  -db Escherichia_coli_str_k_12_substr_mg1655.ASM584v2.pep.all.fa  \
  -query shuffled_lysC.fa \
  <span class="op">&gt;</span> shuffled_lysC_vs_Ecoli_synthesis.tsv

<span class="co">## Check the result</span>
<span class="fu">less</span> shuffled_lysC_vs_Ecoli.txt</code></pre></div>
</div>
<div id="questions-1" class="section level3">
<h3>Questions</h3>
<ol style="list-style-type: decimal">
<li>How many alignments would you expect by chance with the default blastp parameters?</li>
<li>How many alignemnts did you get with the shuffled sequence?</li>
<li>How many of these had an E-value smaller than 1?</li>
<li>How many of these had an E-value smaller than 0.01?</li>
<li>How many of these would you qualify of homologs?</li>
</ol>
</div>
</div>
</div>
<div id="comparing-all-the-proteins-of-a-genome-against-another-genome" class="section level1">
<h1>Comparing all the proteins of a genome against another genome</h1>
<div id="downloading-a-query-genome" class="section level2">
<h2>Downloading a query genome</h2>
<ol style="list-style-type: decimal">
<li><p>In EnsemblGenomes Download page, find the genome of <em>Bacillus subtilis subsp. subtilis str. 168</em>.</p></li>
<li><p>Download the genome annotation (GTF file) and protein sequences (fasta file).</p></li>
</ol>
<div id="solution-4" class="section level3">
<h3>Solution</h3>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## Download genome annotations for query organism</span>
<span class="fu">wget</span> --no-clobber ftp://ftp.ensemblgenomes.org/pub/release-41/bacteria//gtf/bacteria_0_collection/bacillus_subtilis_subsp_subtilis_str_168/Bacillus_subtilis_subsp_subtilis_str_168.ASM904v1.41.gtf.gz

<span class="co">## Download protein sequeces for query organism</span>
<span class="fu">wget</span> --no-clobber ftp://ftp.ensemblgenomes.org/pub/release-41/bacteria//fasta/bacteria_0_collection/bacillus_subtilis_subsp_subtilis_str_168/pep/Bacillus_subtilis_subsp_subtilis_str_168.ASM904v1.pep.all.fa.gz</code></pre></div>
</div>
</div>
<div id="all-proteins-from-e.-coli-versus-e.-coli" class="section level2">
<h2>All proteins from E. coli versus E. coli</h2>
<p><strong>Exercise: </strong> build a command that compares the sequences of all the E.coli proteins against each other, and store the results in a tab-delimited file with a single line per hit (we do not need the alignments).</p>
<hr />
</div>
</div>
<div id="references" class="section level1">
<h1>References</h1>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
